app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventMapper.kt

```diff
@@ -0,0 +1,154 @@
+package org.dhis2.usescases.programEventDetail
+
+import org.dhis2.Bindings.userFriendlyValue
+import org.dhis2.data.tuples.Pair
+import org.dhis2.utils.DateUtils
+import org.hisp.dhis.android.core.D2
+import org.hisp.dhis.android.core.arch.helpers.UidsHelper
+import org.hisp.dhis.android.core.arch.repositories.scope.RepositoryScope
+import org.hisp.dhis.android.core.category.CategoryOptionCombo
+import org.hisp.dhis.android.core.common.State
+import org.hisp.dhis.android.core.event.Event
+import org.hisp.dhis.android.core.program.ProgramStageSection
+import org.hisp.dhis.android.core.trackedentity.TrackedEntityDataValue
+import java.util.Collections
+import java.util.Date
+import javax.inject.Inject
+
+class ProgramEventMapper @Inject constructor(val d2: D2) {
+
+    fun eventToProgramEvent(event: Event): ProgramEventViewModel {
+        val orgUnitName: String = getOrgUnitName(event.organisationUnit()) ?: ""
+        val showInReportsDataElements = mutableListOf<String>()
+        val programStageDataElements = getProgramStageDataElements(event.programStage())
+        programStageDataElements.forEach {
+            if(it.displayInReports() == true) {
+                it.dataElement()?.let { dataElement ->
+                    showInReportsDataElements.add(dataElement.uid())
+                }
+            }
+        }
+        val data = getData(
+            event.trackedEntityDataValues(),
+            showInReportsDataElements,
+            event.programStage()
+        )
+        val hasExpired = isExpired(event)
+        val inOrgUnitRange = checkOrgUnitRange(event.organisationUnit(), event.eventDate()!!)
+        val catOptComb = getCategoryOptionCombo(event.attributeOptionCombo())
+        val attrOptCombo: String =
+            if (catOptComb != null && !catOptComb.displayName().equals("default")) {
+                catOptComb.displayName()!!
+            } else {
+                ""
+            }
```

Comment:
I think you can use ? and remove !! that throws exceptions. Also in your "if" in case displayName is null it can crash

if (!catOptComb?.displayName()?.equals("default")) {
                catOptComb.displayName()
            } else {
                ""
            }

or ternary operator